hera_cal/nucal.py

```diff
@@ -1013,3 +1032,516 @@ def project_u_model_comps_on_spec_axis(u_model_comps, spectral_filters):
+                return_model_comps=True, tol=linear_tol
+            )
+            init_model_comps = project_u_model_comps_on_spec_axis(init_model_comps, self.spectral_filters)
+        else:
+            init_model_comps = fit_nucal_foreground_model(
+                data, data_wgts, self.radial_reds, self.spatial_filters, solver=linear_solver, share_fg_model=share_fg_model, 
+                spectral_filters=self.spectral_filters, return_model_comps=True, tol=linear_tol
+            )
+            
+        # Compute idealized baseline vectors from antenna positions and calibration flags
+        idealized_antpos = abscal._get_idealized_antpos(cal_flags, self.antpos, data.pols())
+
+        if estimate_degeneracies:
+            model = evaluate_foreground_model(
+                self.radial_reds, init_model_comps, spatial_filters=self.spatial_filters, spectral_filters=self.spectral_filters
+            )
+            amplitude, tip_tilt = self._estimate_degeneracies(data, model, data_wgts)
+        else:
+            amplitude = {pol: np.ones((data.shape)) for pol in data.pols()}
+            tip_tilt = {
+                pol: np.zeros((idealized_antpos[list(idealized_antpos.keys())[0]].shape[0], data.shape[0], data.shape[1])) 
+                for pol in data.pols()
+            }
+        
+        # Initialize model parameters and metadata dictionaries for storing results
+        model_parameters = {}
+        metadata = {}
+
+        for pol in data.pols():
+            # Initialize model parameters
+            init_model_parameters = {}
+
+            # Separate data into real and imaginary components
+            data_real = jnp.array([
+                data[blkey].real for rdgrp in self.radial_reds.get_pol(pol) for blkey in rdgrp
+            ])
+            data_imag = jnp.array([
+                data[blkey].imag for rdgrp in self.radial_reds.get_pol(pol) for blkey in rdgrp
+            ])
+
+            # Initialize model parameters
+            init_model_parameters = {
+                "fg_r": [init_model_comps[rdgrp[0]].real for rdgrp in self.radial_reds.get_pol(pol)],
+                "fg_i": [init_model_comps[rdgrp[0]].imag for rdgrp in self.radial_reds.get_pol(pol)],
+                "amplitude": amplitude[pol],
+                "tip_tilt": tip_tilt[pol]
+            }
+            # unpack wgts/filters
+            wgts = jnp.array([
+                data_wgts[blkey] for rdgrp in self.radial_reds.get_pol(pol) for blkey in rdgrp
+            ])
+
+            # Set spectral filters 
+            spatial_filters = [
+                jnp.array([self.spatial_filters[blkey] for blkey in rdgrp]) 
+                for rdgrp in self.radial_reds.get_pol(pol)
+            ]
+
+            # Compute idealized baseline vectors
+            idealized_blvecs = np.array([
+                idealized_antpos[blkey[1]] - idealized_antpos[blkey[0]]
+                for rdgrp in self.radial_reds.get_pol(pol) for blkey in rdgrp
+            ])
```

Comment:
This feels like a lot of copy-pasted code. Can it be simplified?

Reply:
The data, weights, and filters need to be organized in jax arrays, but I was hoping to prevent copying the filters and data for both pol at the same time to reduce memory usage. If you feel that it should be changed I can, but this seemed like the simplest way without copying too much of the data